### DISCUSSION TITLE
ARC-0020 - token standard allowing public/private transfer to contracts

### DISCUSSION THREAD

**Discussion by r001** [ProvableHQ/ARCs#59] [ARCs]
Let's list the token standard proposals so far:

1.  [PULL#14](https://github.com/AleoHQ/ARCs/pull/14) ([README.md](https://github.com/lambdaclass/aleo_ARCs/blob/arc-0020/arc-0020/README.md), [main.aleo](https://github.com/lambdaclass/ARC20_leo/blob/main/src/main.leo), [discussion](https://github.com/AleoHQ/ARCs/discussions/13)) has listed the required instructions for token standard that allows **contracts** to interact with tokens **publicly**:
     They are: **balance_of**, **total_supply**, **approve**, **transfer_from**, **transfer**, and **allowance**.
     ARC20 includes 3 optional rules. They are: **decimals**, **name**, **symbol**.
     Advantages:
     1. Provides methods for public contract interaction.
     Disadvantages:
     1. Had to use very old snarkVM features.

     Thanks to [@EdVerally](https://github.com/EdVeralli) for his contribution.  

2. [PULL#34](https://github.com/AleoHQ/ARCs/pull/34)  ([README.md](https://github.com/ghostant-1017/ARCs/blob/feat/arc20/arc-0020/README.md), (main.aleo - in README.md), [discussion](https://github.com/AleoHQ/ARCs/discussions/35)) created a unique standard by using records even for public transfers. 
     Advantages: 
     1. It also allows for upgradability. 
     3. It allows to handle unlimited tokens with one single codebase, which is an advantage of this method (with [HarukaMa](https://github.com/HarukaMa)'s [concerns](https://github.com/AleoHQ/ARCs/pull/34#issuecomment-1653911564))
     
     Disadvantages:
     1. Does not provide solution for sending tokens to contracts.
     3. Credential concept is strange.
     
     Thanks to [ghostant-1017](https://github.com/ghostant-1017) for this unique addition.

3.  [PULL#41](https://github.com/AleoHQ/ARCs/pull/41) ([README.md](https://github.com/vicsn/ARCs/blob/arc20_approvals/arc-0020/README.md), [main.aleo](https://github.com/vicsn/ARCs/blob/arc20_approvals/arc-0020/token.aleo), [discussion](https://github.com/AleoHQ/ARCs/discussions/42))
     Advantages:
     1. Has a solution for public contract usage. 
     2. Clear and usable.
     
     Disadvantages:
     1. Applies the ([discussion](https://github.com/AleoHQ/ARCs/pull/41#issuecomment-1851018566)) `approve()` transition, instead of offline signatures, which is both more awkward, uses more system resources at each calling of `approve()`, and the codebase is also around 41 lines longer than that of offline signatures.
     
     Thanks for the Aleo team to come up with this solution.
 
None of the above concepts address the processing of private tokens by smart contracts. 

4. Current ARC0020 **[here](https://github.com/r001/ARCs/blob/master/arc-0020/README.md)**. ([README.md](https://github.com/AleoHQ/ARCs/pull/41#issuecomment-1851018566), [main.aleo](https://github.com/r001/arc20_0001/blob/main/reference/main.aleo), [discussion]())
     Advantages:
     1. Provides offline signature solution for [public tokens to  be used by contracts](https://github.com/r001/ARCs/blob/master/arc-0020/README.md#2297-transfer-tokens-to-contracts-privately---must-be-implmented).
     2. Provides solution for [private tokens to be used by contracts](https://github.com/r001/ARCs/blob/master/arc-0020/README.md#2_3_1_deposit_private_token_record_to_contract).
     3. Provides simple solution to make sure [contract is indeed belonging to the company website](https://github.com/r001/ARCs/blob/master/arc-0020/README.md#2_2_3_company_signature_must_be_implemented).
     
     Disadvantages:
     1. [No solution for](https://github.com/r001/ARCs/blob/master/arc-0020/README.md#2_2_10_total_supply_transition_not_implementable) `total_supply`.
     2. Does not create one single contract for all tokens (would save storage space). 
     3. Current implementation of [ZERO_ADDRESS](https://github.com/r001/ARCs/blob/master/arc-0020/README.md#2112-zero_address) is awkward a bit.

**Comment by zklim** [comment]
The offline signing approval method is currently vulnerable to replay attack as the struct hash is not constrained enough to include information like network.id, function id, program id and nonce etc.

**Reply by zklim** [reply]
The current hash only takes `to`, `from`, `amount` and `expiry` into account. The `to` address can use this signature to drain `amount` many times from `from` address as long it has balance and not expired.

Not to mention the `to` address can replay this signature on other tokens on the same / other network that implemented the exact same signature scheme.
